ext/oj/dump_compat.c

```diff
@@ -645,6 +645,16 @@ dump_float(VALUE obj, int depth, Out out, bool as_ok) {
     *out->cur = '\0';
 }
 
+static void
+dump_time(VALUE obj, int depth, Out out, bool as_ok) {
+    if ((as_ok && rb_respond_to(obj, oj_to_json_id)) &&
+        (Yes == out->opts->to_json || rb_respond_to(obj, oj_as_json_id))) {
```

Comment:
Since `#to_json` is going to be called, there is no need to check for `#as_json`.  Maybe something like this would be better:

```c
if (as_ok && rb_respond_to(obj, oj_to_json_id)) {
```

Reply:
https://github.com/ohler55/oj/blob/38be7638f2933b123d4e67ac8f40da06617e346b/ext/oj/dump_compat.c#L699

After executed above line, `as_ok` will always have `true`.
Therefore, I add  #as_json check.

Reply:
And, if it don't call #to_json when #as_json exists, the test in oj will fail....

Reply:
`#to_json` is so slow, so I don't want to call it if possible.

Reply:
Right, it is slow but it doesn't make sense to check for `#as_json` and then call `#to_json`.

Logic to call `#as_json` instead would be ok but the respond_to check should match the funcall.

Reply:
If `Time#to_json` is overridden by `require 'json/add/time'`, it need to call `Time#to_json` because behavior is changed.

I want to know a way whether the method is overridden.... but...
So, this PR is checking that `Time#as_json` is added by `require 'json/add/time'` (because I think it can be assumed that `Time#to_json` has been overridden by `require 'json/add/time'`)

Reply:
Is there a more direct way to check? By keying off of `#as_json` any other gem or use code that added `#as_json` would match as well. Is there a way to determine where a method was defined? Maybe checking the method object or the method `#hash`?

Reply:
I wonder if the best approach might be to monkey patch time as part of mimic_JSON. What do you think?

Reply:
I tried to simplify the branching by calling as_json beforehand.
https://github.com/ohler55/oj/pull/694/commits/127b5a3ae03c7eb896423c823b24b73f647efd43

Reply:
hmm, fails the unit test...

> I wonder if the best approach might be to monkey patch time as part of mimic_JSON.

I will consider your approach.